File: article-api/src/main/kotlin/com/sns/article/reaction/domain/Reaction.java
package com.sns.article.reaction.domain;

import org.springframework.context.ApplicationEventPublisher;
import org.springframework.data.annotation.Id;
import org.springframework.data.relational.core.mapping.Embedded;

public class Reaction {
	@Id
	private final Long id;

	@Embedded(onEmpty = Embedded.OnEmpty.USE_NULL)
	private final ReactionTarget target;

	private final ReactionType type;

	private final String userId;

	public Reaction(Long id, ReactionTarget target, ReactionType type, String userId) {
		this.id = id;
		this.target = target;
		this.type = type;
		this.userId = userId;
	}

	public Long getId() {
		return id;
	}

	public ReactionTarget getTarget() {
		return target;
	}

	public ReactionType getType() {
		return type;
	}

	public String getUserId() {
		return userId;
	}

	public void created(ApplicationEventPublisher eventPublisher) {
		if (eventPublisher != null) {
			eventPublisher.publishEvent(new ReactionCreatedEvent(id));
		}
	}
}

File: article-api/src/main/kotlin/com/sns/article/component/reaction/domains/Reaction.kt
package com.sns.article.component.reaction.domains

import com.sns.commons.exceptions.NoAuthorityException
import org.springframework.context.ApplicationEventPublisher
import org.springframework.data.annotation.Id
import org.springframework.data.relational.core.mapping.Column
import org.springframework.data.relational.core.mapping.Embedded
import org.springframework.data.relational.core.mapping.Table

@Table(value = "reaction")
class Reaction(
    @field:Id @field:Column(value = "id") var id: Long? = null,
    @field:Embedded(onEmpty = Embedded.OnEmpty.USE_EMPTY) val target: ReactionTarget,
    @field:Column(value = "type") val type: ReactionType,
    @field:Column(value = "user_id") val userId: String
) {
    fun created(eventPublisher: ApplicationEventPublisher?) {
        eventPublisher?.publishEvent(ReactionCreatedEvent(id!!))
    }

    fun deleteBy(userId: String) {
        if (userId != userId) {
            throw NoAuthorityException()
        }
    }
}
